include/onnxruntime/core/session/onnxruntime_c_api.h

```diff
@@ -836,6 +836,29 @@ struct OrtApi {
    */
   ORT_API2_STATUS(ReleaseAvailableProviders, _In_ char **ptr,
                   _In_ int providers_length);
+
+  /**
+     * \param value A tensor created from OrtCreateTensor... function.
+     * \param len element length, not including the trailing '\0' chars.
+     * \param index index of string tensor element, length of element at index will be returned.
+     */
+  ORT_API2_STATUS(GetStringTensorElementLength, _In_ const OrtValue* value, _Out_ size_t* out, size_t index);
+
+  /**
+     * \param s string element contents. The string is NOT null-terminated.
+     * \param value A tensor created from OrtCreateTensor... function.
+     * \param s_len element length, get it from OrtGetStringTensorElementLength.
+     * \param index offset of element of tensor to return.
+     */
+  ORT_API2_STATUS(GetStringTensorElement, _In_ const OrtValue* value, _Out_writes_bytes_all_(s_len) void* s,
+                  size_t s_len, size_t index);
+
+  /**
+     * \param value A tensor created from OrtCreateTensor... function.
+     * \param s A null terminated.
+     * \param index index of string tensor element to fill 
+     */
+  ORT_API2_STATUS(FillStringTensorElement, _Inout_ OrtValue* value, _In_ const char* s, size_t index);
```

Comment:
I didn't look too deeply, but is it possible to require a length of the string passed in vs using null terminated?

Reply:
Another possible issue I thought of with adding this API is it assumes we can easily change a string in the middle of a tensor quickly. If we changed our internal implementation to pack the strings into a single block of memory (as std::string is bad for locality and performance of bulk strings) this would now be very inefficient. We could probably just say it's slow in the future if we change the internals, but just wanted to mention it now.

Reply:
1. Regarding the first comment:   In the implementation, the char* is assigned to the tensor using the string assignment operator. As I understand, passing in a non-null terminated char array to a string assignment operator would result in undefined behavior. Here's the code for context: 

```auto* dst = tensor->MutableData<std::string>();
.....

  dst[index] = s;
```  

Would you have any alternative suggestion? 

2.  For the second comment: I totally agree. I'll make a note of this in our design document. It's definitely good to understand the trade-offs here between the flexibility of adding strings in the middle of the tensor and potential future performance loss.

Reply:
There is a std::string assignment that lets you specify a length. Given we use null terminated strings elsewhere, it does make sense to not require it here. It just limits the users to null terminated strings (and the length must be calculated when it's used). So it's okay to leave as is.

Reply:
Got it, thanks for clarifying!